src/main/java/com/gotocompany/depot/maxcompute/record/ProtoDataColumnRecordDecorator.java

```diff
@@ -0,0 +1,79 @@
+package com.gotocompany.depot.maxcompute.record;
+
+import com.google.protobuf.Descriptors;
+import com.gotocompany.depot.config.SinkConfig;
+import com.gotocompany.depot.maxcompute.converter.ConverterOrchestrator;
+import com.gotocompany.depot.maxcompute.model.RecordWrapper;
+import com.gotocompany.depot.maxcompute.schema.partition.DefaultPartitioningStrategy;
+import com.gotocompany.depot.maxcompute.schema.partition.PartitioningStrategy;
+import com.gotocompany.depot.maxcompute.schema.partition.TimestampPartitioningStrategy;
+import com.gotocompany.depot.message.Message;
+import com.gotocompany.depot.message.MessageParser;
+import com.gotocompany.depot.message.ParsedMessage;
+import com.gotocompany.depot.message.SinkConnectorSchemaMessageMode;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Optional;
+
+public class ProtoDataColumnRecordDecorator extends RecordDecorator {
+
+    private final ConverterOrchestrator converterOrchestrator;
+    private final MessageParser protoMessageParser;
+    private final PartitioningStrategy partitioningStrategy;
+    private final SinkConfig sinkConfig;
+
+    public ProtoDataColumnRecordDecorator(RecordDecorator decorator,
+                                          ConverterOrchestrator converterOrchestrator,
+                                          MessageParser messageParser,
+                                          SinkConfig sinkConfig,
+                                          PartitioningStrategy partitioningStrategy) {
+        super(decorator);
+        this.converterOrchestrator = converterOrchestrator;
+        this.protoMessageParser = messageParser;
+        this.partitioningStrategy = partitioningStrategy;
+        this.sinkConfig = sinkConfig;
+    }
+
+    @Override
+    public void append(RecordWrapper recordWrapper, Message message) throws IOException {
+        String schemaClass = getSchemaClass();
+        ParsedMessage parsedMessage = protoMessageParser.parse(message, sinkConfig.getSinkConnectorSchemaMessageMode(), schemaClass);
+        parsedMessage.validate(sinkConfig);
+        com.google.protobuf.Message protoMessage = (com.google.protobuf.Message) parsedMessage.getRaw();
+        Map<Descriptors.FieldDescriptor, Object> fields = protoMessage.getAllFields();
+        for (Map.Entry<Descriptors.FieldDescriptor, Object> entry : fields.entrySet()) {
+            if (entry.getKey().getName().equals(getPartitionFieldName()) && shouldReplaceOriginalColumn()) {
+                continue;
+            }
+            recordWrapper.getRecord()
+                    .set(entry.getKey().getName(), converterOrchestrator.convert(entry.getKey(), entry.getValue()));
+        }
```

Comment:
No handling for unknown proto fields or incompatible field types. Missing validation for required fields. Should add proper schema validation and handle unknown/deprecated fields.

Reply:
- Required field validation is handled during deserialization, it should throw error for non existent required field. Do we have this kind of validation in BQ Sink as well?
- Line 42 takes care of handling unknown field.
- Wdym by incompatible field types